### PR TITLE
Make parallel::merge is stable. (Fix #2964.)

### LINKED ISSUE #2964: parallel merge is not stable
[merge_testcase.zip](https://github.com/STEllAR-GROUP/hpx/files/1392241/merge_testcase.zip)
The documentation for the merge algorithm states _For equivalent elements in the original two ranges, the elements from the first range precede the elements from the second range_.  I find that this is sometimes not the case when the parallel execution policy is chosen.

The attached testcase merges two ranges of <int, char> pairs, ordered by the first value.  For my compiler (gcc 6.3.0) I find that the resulting merged sequence interleaves equivalent elements from the input ranges, instead of using all of the equivalent elements from the first range first.  Specifically, when merging `(3,a), (3,b)` with `(3,c)` the result is `(3,a), (3,c), (3,b)`.

Due to the high threshold (65536) for using the parallel version of the algorithm, reproducing this requires modifying [this line](https://github.com/STEllAR-GROUP/hpx/blob/master/hpx/parallel/algorithms/merge.hpp#L173) so the threshold is something low (I used 10).

### PR DESCRIPTION
**Pull request by taeguk**

### CODE REVIEW

File: tests/regressions/parallel/stable_merge_2964.cpp [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+          {
+              return std::get<0>(a) < std::get<0>(b);
+          });
+
+    std::for_each(result.begin(), result.end(),
+        [](auto const& a)
+        {
+            std::cout << "(" << std::get<0>(a) << ", " << std::get<1>(a) << ") ";
+        });
+    std::cout << "\n";
```
Comment (sithhell) [inline tests/regressions/parallel/stable_merge_2964.cpp]: Could we actually test something here?

File: tests/regressions/parallel/stable_merge_2964.cpp [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+          [](auto const& a, auto const& b)
+          {
+              return std::get<0>(a) < std::get<0>(b);
+          });
+
+    std::for_each(result.begin(), result.end(),
+        [](auto const& a)
+        {
+            std::cout << "(" << std::get<0>(a) << ", " << std::get<1>(a) << ") ";
+        });
```
Comment (sithhell) [inline tests/regressions/parallel/stable_merge_2964.cpp]: Same here.

File: tests/regressions/parallel/stable_merge_2964.cpp (hunk 1 above)
Comment (taeguk) [reply tests/regressions/parallel/stable_merge_2964.cpp]: @sithhell I performed test with temporarily setting threshold to 10. And I got correct results.
Of course, by the source code itself, it seems that we actually can test nothing.
Before we test, we must set threshold to 10 temporarily.

Comment (taeguk) [reply tests/regressions/parallel/stable_merge_2964.cpp]: @sithhell To perform a complete test with the source code itself without any temporary modification, we need to be able to control threshold value without editing the source code. But, in now, we can't.

**Comment by sithhell** [comment]
You completely misunderstood. That the test isn't really testing the
implementation is one problem. What I referred to here is that the test
only outputs stuff instead of reporting potential problems, see the other
tests including our lightweight_test.hpp header.

**Comment by taeguk** [comment]
@sithhell Yes, you're right. I modified it as you said.

File: tests/regressions/parallel/stable_merge_2964.cpp [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+        a1.begin(), a1.end(),
+        solution.begin(),
+        [](auto const& a, auto const& b)
+        {
+            return std::get<0>(a) < std::get<0>(b);
+        });
+    
+    equality = std::equal(
+        result.begin(), result.end(),
+        solution.begin(), solution.end());
```
Comment (sithhell) [inline tests/regressions/parallel/stable_merge_2964.cpp]: Wouldn't it be easy enough to extend the test to be run on a number of elements that exceed the threshold?

Comment (taeguk) [reply tests/regressions/parallel/stable_merge_2964.cpp]: @sithhell I did that in my PR which I'm preparing. I added that to unit test.

Comment (hkaiser) [reply tests/regressions/parallel/stable_merge_2964.cpp]: No need to have it twice.

Comment (taeguk) [reply tests/regressions/parallel/stable_merge_2964.cpp]: @sithhell @hkaiser I added test code with a enough number of elements which is filled randomly for both regression and unit tests.

**Comment by hkaiser** [comment]
Thanks @taeguk, LGTM now.

**Comment by hkaiser** [comment]
The remaining inspect error is unrelated and was already fixed on master.

**Pull request merged**